Support for cusparselt 0.6.0 adapted from https://github.com/cupy/cupy/pull/8074
but with bug fixed for compressed size functions that were wrong.
author: micketeer@gmail.com
diff -ru cupy_backends.orig/cuda/libs/cusparselt.pxd cupy_backends/cuda/libs/cusparselt.pxd
--- cupy_backends.orig/cuda/libs/cusparselt.pxd	2024-03-23 17:08:35.676189389 +0100
+++ cupy_backends/cuda/libs/cusparselt.pxd	2024-03-23 17:30:34.931591841 +0100
@@ -13,18 +13,20 @@
     CUSPARSELT_MAT_BATCH_STRIDE = 1  # READ/WRITE
 
     # cusparseComputeType
-    CUSPARSE_COMPUTE_16F = 0
-    CUSPARSE_COMPUTE_32I = 1
-    CUSPARSE_COMPUTE_TF32 = 2
-    CUSPARSE_COMPUTE_TF32_FAST = 3
+    CUSPARSE_COMPUTE_32I = 0
+    CUSPARSE_COMPUTE_16F = 1
+    CUSPARSE_COMPUTE_32F = 2
 
     # cusparseLtMatmulDescAttribute_t
     CUSPARSELT_MATMUL_ACTIVATION_RELU = 0             # READ/WRITE
     CUSPARSELT_MATMUL_ACTIVATION_RELU_UPPERBOUND = 1  # READ/WRITE
     CUSPARSELT_MATMUL_ACTIVATION_RELU_THRESHOLD = 2   # READ/WRITE
     CUSPARSELT_MATMUL_ACTIVATION_GELU = 3             # READ/WRITE
-    CUSPARSELT_MATMUL_BIAS_STRIDE = 4                 # READ/WRITE
-    CUSPARSELT_MATMUL_BIAS_POINTER = 5                # READ/WRITE
+    CUSPARSELT_MATMUL_ACTIVATION_GELU_SCALING = 4     # READ/WRITE
+    CUSPARSELT_MATMUL_ALPHA_VECTOR_SCALING = 5        # READ/WRITE
+    CUSPARSELT_MATMUL_BETA_VECTOR_SCALING = 6         # READ/WRITE
+    CUSPARSELT_MATMUL_BIAS_POINTER = 7                # READ/WRITE
+    CUSPARSELT_MATMUL_BIAS_STRIDE = 8                 # READ/WRITE
 
     # cusparseLtMatmulAlg_t
     CUSPARSELT_MATMUL_ALG_DEFAULT = 0
@@ -33,6 +35,14 @@
     CUSPARSELT_MATMUL_ALG_CONFIG_ID = 0      # NOQA, READ/WRITE
     CUSPARSELT_MATMUL_ALG_CONFIG_MAX_ID = 1  # NOQA, READ-ONLY
     CUSPARSELT_MATMUL_SEARCH_ITERATIONS = 2  # NOQA, READ/WRITE
+    CUSPARSELT_MATMUL_SPLIT_K = 3            # NOQA, READ/WRITE
+    CUSPARSELT_MATMUL_SPLIT_K_MODE =4        # NOQA, READ/WRITE
+    CUSPARSELT_MATMUL_SPLIT_K_BUFFERS=5      # NOQA, READ/WRITE
+
+    # cusparseLtSplitKMode_t
+    CUSPARSELT_INVALID_MODE = 0
+    CUSPARSELT_SPLIT_K_MODE_ONE_KERNEL  = 1
+    CUSPARSELT_SPLIT_K_MODE_TWO_KERNELS = 2
 
     # cusparseLtPruneAlg_t
     CUSPARSELT_PRUNE_SPMMA_TILE = 0
diff -ru cupy_backends.orig/cuda/libs/cusparselt.pyx cupy_backends/cuda/libs/cusparselt.pyx
--- cupy_backends.orig/cuda/libs/cusparselt.pyx	2024-03-23 17:08:35.676189389 +0100
+++ cupy_backends/cuda/libs/cusparselt.pyx	2024-03-23 18:03:43.706215397 +0100
@@ -11,6 +11,13 @@
 from cupy_backends.cuda.libs import cusparse as _cusparse
 
 
+###############################################################################
+# Types
+###############################################################################
+cdef extern from *:
+    ctypedef void* LibraryPropertyType 'libraryPropertyType_t'
+
+
 cdef extern from '../../cupy_cusparselt.h' nogil:
     ctypedef int cusparseStatus_t 'cusparseStatus_t'
     ctypedef int cusparseOrder_t 'cusparseOrder_t'
@@ -36,11 +43,14 @@
     ctypedef int cusparseOperation_t 'cusparseOperation_t'
     ctypedef int cusparseLtMatmulAlg_t 'cusparseLtMatmulAlg_t'
     ctypedef int cusparseLtMatmulAlgAttribute_t 'cusparseLtMatmulAlgAttribute_t'  # NOQA
+    ctypedef int cusparseLtSplitKMode_t 'cusparseLtSplitKMode_t'
     ctypedef int cusparseLtPruneAlg_t 'cusparseLtPruneAlg_t'
 
     # Management Functions
     cusparseStatus_t cusparseLtInit(cusparseLtHandle_t* handle)
     cusparseStatus_t cusparseLtDestroy(const cusparseLtHandle_t* handle)
+    cusparseStatus_t cusparseLtGetVersion(const cusparseLtHandle_t* handle, int* version)
+    cusparseStatus_t cusparseLtGetProperty(LibraryPropertyType propertyType, int* value)
 
     # Matmul Functions
     cusparseStatus_t cusparseLtDenseDescriptorInit(
@@ -66,6 +76,11 @@
         const cusparseLtMatDescriptor_t* matDescr,
         cusparseLtMatDescAttribute_t matAttribute,
         void* data, size_t dataSize)
+    cusparseStatus_t cusparseLtMatDescSetAttribute(
+        const cusparseLtHandle_t* handle,
+        const cusparseLtMatDescriptor_t* matDescr,
+        cusparseLtMatDescAttribute_t matAttribute,
+        void* data, size_t dataSize)
     cusparseStatus_t cusparseLtMatmulDescriptorInit(
         const cusparseLtHandle_t* handle,
         cusparseLtMatmulDescriptor_t* matMulDescr,
@@ -95,17 +110,21 @@
         const cusparseLtHandle_t* handle,
         cusparseLtMatmulAlgSelection_t* algSelection,
         cusparseLtMatmulAlgAttribute_t attribute,
-        const void* data, size_t ataSize)
-    cusparseStatus_t cusparseLtMatmulGetWorkspace(
+        const void* data, size_t dataSize)
+    cusparseStatus_t cusparseLtMatmulAlgGetAttribute(
         const cusparseLtHandle_t* handle,
         const cusparseLtMatmulAlgSelection_t* algSelection,
+        cusparseLtMatmulAlgAttribute_t attribute,
+        void* data, size_t dataSize)
+    cusparseStatus_t cusparseLtMatmulGetWorkspace(
+        const cusparseLtHandle_t* handle,
+        const cusparseLtMatmulPlan_t* plan,
         size_t* workspaceSize)
     cusparseStatus_t cusparseLtMatmulPlanInit(
         const cusparseLtHandle_t* handle,
         cusparseLtMatmulPlan_t* plan,
         const cusparseLtMatmulDescriptor_t* matmulDescr,
-        const cusparseLtMatmulAlgSelection_t* algSelection,
-        size_t workspaceSize)
+        const cusparseLtMatmulAlgSelection_t* algSelection)
     cusparseStatus_t cusparseLtMatmulPlanDestroy(
         const cusparseLtMatmulPlan_t* plan)
     cusparseStatus_t cusparseLtMatmul(
@@ -113,6 +132,11 @@
         const void* alpha, const void* d_A, const void* d_B,
         const void* beta, const void* d_C, void* d_D,
         void* workspace, runtime.Stream* streams, int32_t numStreams)
+    cusparseStatus_t cusparseLtMatmulSearch(
+        const cusparseLtHandle_t* handle, cusparseLtMatmulPlan_t* plan,
+        const void* alpha, const void* d_A, const void* d_B,
+        const void* beta, const void* d_C, void* d_D,
+        void* workspace, runtime.Stream* streams, int32_t numStreams)
 
     # Helper Functions
     cusparseStatus_t cusparseLtSpMMAPrune(
@@ -123,7 +147,7 @@
     cusparseStatus_t cusparseLtSpMMAPruneCheck(
         const cusparseLtHandle_t* handle,
         const cusparseLtMatmulDescriptor_t* matmulDescr,
-        const void* d_in, int* valid, runtime.Stream stream)
+        const void* d_in, int* d_valid, runtime.Stream stream)
     cusparseStatus_t cusparseLtSpMMAPrune2(
         const cusparseLtHandle_t* handle,
         const cusparseLtMatDescriptor_t* sparseMatDescr,
@@ -136,19 +160,22 @@
         runtime.Stream stream)
     cusparseStatus_t cusparseLtSpMMACompressedSize(
         const cusparseLtHandle_t* handle, const cusparseLtMatmulPlan_t* plan,
-        size_t* compressedSize)
+        size_t* compressedSize,
+        size_t* compressedBufferSize)
     cusparseStatus_t cusparseLtSpMMACompress(
         const cusparseLtHandle_t* handle, const cusparseLtMatmulPlan_t* plan,
-        const void* d_dense, void* d_compressed, runtime.Stream stream)
+        const void* d_dense, void* d_compressed, void* d_compressed_buffer,
+        runtime.Stream stream)
     cusparseStatus_t cusparseLtSpMMACompressedSize2(
         const cusparseLtHandle_t* handle,
         const cusparseLtMatDescriptor_t* sparseMatDescr,
-        size_t* compressedSize)
+        size_t* compressedSize,
+        size_t* compressedBufferSize)
     cusparseStatus_t cusparseLtSpMMACompress2(
         const cusparseLtHandle_t* handle,
         const cusparseLtMatDescriptor_t* sparseMatDescr,
         int isSparseA, cusparseOperation_t op, const void* d_dense,
-        void* d_compressed, runtime.Stream stream)
+        void* d_compressed, void* d_compressed_buffer, runtime.Stream stream)
 
     # Build-time version
     int CUSPARSELT_VERSION
@@ -370,28 +397,36 @@
         <const void*> data, <size_t> dataSize)
     check_status(status)
 
+cpdef matmulAlgGetAttribute(Handle handle, MatmulAlgSelection algSelection,
+                            attribute, size_t data, size_t dataSize):
+    """Gets the attribute related to algorithm selection descriptor."""
+    status = cusparseLtMatmulAlgGetAttribute(
+        <const cusparseLtHandle_t*> handle._ptr,
+        <const cusparseLtMatmulAlgSelection_t*> algSelection._ptr,
+        <cusparseLtMatmulAlgAttribute_t> attribute,
+        <void*> data, <size_t> dataSize)
+    check_status(status)
+
 cpdef size_t matmulGetWorkspace(Handle handle,
-                                MatmulAlgSelection algSelection):
+                                MatmulPlan plan):
     """Determines the required workspace size"""
     cdef size_t workspaceSize
     status = cusparseLtMatmulGetWorkspace(
         <const cusparseLtHandle_t*> handle._ptr,
-        <const cusparseLtMatmulAlgSelection_t*> algSelection._ptr,
+        <const cusparseLtMatmulPlan_t*> plan._ptr,
         &workspaceSize)
     check_status(status)
     return workspaceSize
 
 cpdef matmulPlanInit(Handle handle, MatmulPlan plan,
                      MatmulDescriptor matmulDescr,
-                     MatmulAlgSelection algSelection,
-                     size_t workspaceSize):
+                     MatmulAlgSelection algSelection):
     """Initializes the plan."""
     status = cusparseLtMatmulPlanInit(
         <const cusparseLtHandle_t*> handle._ptr,
         <cusparseLtMatmulPlan_t*> plan._ptr,
         <const cusparseLtMatmulDescriptor_t*> matmulDescr._ptr,
-        <const cusparseLtMatmulAlgSelection_t*> algSelection._ptr,
-        <size_t> workspaceSize)
+        <const cusparseLtMatmulAlgSelection_t*> algSelection._ptr)
     check_status(status)
 
 cpdef matmulPlanDestroy(MatmulPlan plan):
@@ -412,6 +447,18 @@
         <void*> workspace, <runtime.Stream*> NULL, <int32_t> 0)
     check_status(status)
 
+cpdef matmulSearch(Handle handle, MatmulPlan plan,
+             size_t alpha, size_t d_A, size_t d_B,
+             size_t beta, size_t d_C, size_t d_D, size_t workspace):
+    """Evaluates all available algorithms for the matrix multiplication"""
+    status = cusparseLtMatmulSearch(
+        <const cusparseLtHandle_t*> handle._ptr,
+        <cusparseLtMatmulPlan_t*> plan._ptr,
+        <const void*> alpha, <const void*> d_A, <const void*> d_B,
+        <const void*> beta, <const void*> d_C, <void*> d_D,
+        <void*> workspace, <runtime.Stream*> NULL, <int32_t> 0)
+    check_status(status)
+
 ###############################################################################
 # cuSPARSELt: Helper Functions
 ###############################################################################
@@ -428,13 +475,13 @@
     check_status(status)
 
 cpdef spMMAPruneCheck(Handle handle, MatmulDescriptor matmulDescr,
-                      size_t d_in, size_t valid):
+                      size_t d_in, size_t d_valid):
     """Checks the correctness of the pruning structure"""
     cdef intptr_t stream = stream_module.get_current_stream_ptr()
     status = cusparseLtSpMMAPruneCheck(
         <const cusparseLtHandle_t*> handle._ptr,
         <const cusparseLtMatmulDescriptor_t*> matmulDescr._ptr,
-        <const void*> d_in, <int*> valid, <runtime.Stream> stream)
+        <const void*> d_in, <int*> d_valid, <runtime.Stream> stream)
     check_status(status)
 
 cpdef spMMAPrune2(Handle handle, MatDescriptor sparseMatDescr, isSparseA,
@@ -460,47 +507,52 @@
         <int*> d_valid, <runtime.Stream> stream)
     check_status(status)
 
-cpdef size_t spMMACompressedSize(Handle handle, MatmulPlan plan):
+cpdef spMMACompressedSize(Handle handle, MatmulPlan plan):
     """Provides the size of the compressed matrix"""
     cdef size_t compressedSize
+    cdef size_t compressedBufferSize
     status = cusparseLtSpMMACompressedSize(
         <const cusparseLtHandle_t*> handle._ptr,
         <const cusparseLtMatmulPlan_t*> plan._ptr,
-        &compressedSize)
+        &compressedSize, &compressedBufferSize)
     check_status(status)
-    return compressedSize
+    return compressedSize, compressedBufferSize
 
 cpdef spMMACompress(Handle handle, MatmulPlan plan,
-                    size_t d_dense, size_t d_compressed):
+                    size_t d_dense, size_t d_compressed,
+                    size_t d_compressed_buffer):
     """Compresses a dense matrix d_dense."""
     cdef intptr_t stream = stream_module.get_current_stream_ptr()
     status = cusparseLtSpMMACompress(
         <const cusparseLtHandle_t*> handle._ptr,
         <const cusparseLtMatmulPlan_t*> plan._ptr,
-        <const void*> d_dense, <void*> d_compressed, <runtime.Stream> stream)
+        <const void*> d_dense, <void*> d_compressed,
+        <void*> d_compressed_buffer, <runtime.Stream> stream)
     check_status(status)
 
-cpdef size_t spMMACompressedSize2(Handle handle, MatDescriptor sparseMatDescr):
+cpdef spMMACompressedSize2(Handle handle, MatDescriptor sparseMatDescr):
     """Provides the size of the compressed matrix"""
     cdef size_t compressedSize
+    cdef size_t compressedBufferSize
     status = cusparseLtSpMMACompressedSize2(
         <const cusparseLtHandle_t*> handle._ptr,
         <const cusparseLtMatDescriptor_t*> sparseMatDescr._ptr,
-        &compressedSize)
+        &compressedSize, &compressedBufferSize)
     check_status(status)
-    return compressedSize
+    return compressedSize, compressedBufferSize
 
 cpdef spMMACompress2(Handle handle, MatDescriptor sparseMatDescr,
-                     isSparseA, op, size_t d_dense, size_t d_compressed):
+                     isSparseA, op, size_t d_dense,
+                     size_t d_compressed, size_t d_compressed_buffer):
     """Compresses a dense matrix d_dense."""
     cdef intptr_t stream = stream_module.get_current_stream_ptr()
     status = cusparseLtSpMMACompress2(
         <const cusparseLtHandle_t*> handle._ptr,
         <const cusparseLtMatDescriptor_t*> sparseMatDescr._ptr,
         <int> isSparseA, <cusparseOperation_t> op, <const void*> d_dense,
-        <void*> d_compressed, <runtime.Stream> stream)
+        <void*> d_compressed, <void*> d_compressed_buffer,
+        <runtime.Stream> stream)
     check_status(status)
 
-
 def get_build_version():
     return CUSPARSELT_VERSION
diff -ru cupy_backends.orig/stub/cupy_cusparselt.h cupy_backends/stub/cupy_cusparselt.h
--- cupy_backends.orig/stub/cupy_cusparselt.h	2024-03-23 17:08:35.677189397 +0100
+++ cupy_backends/stub/cupy_cusparselt.h	2024-03-23 18:15:04.734544642 +0100
@@ -19,6 +19,7 @@
 typedef enum {} cusparseLtMatmulDescAttribute_t;
 typedef enum {} cusparseLtMatmulAlg_t;
 typedef enum {} cusparseLtMatmulAlgAttribute_t;
+typedef enum {} cusparseLtSplitKMode_t;
 typedef enum {} cusparseLtPruneAlg_t;
 
 typedef void* cudaStream_t;
@@ -36,6 +37,14 @@
     return CUSPARSE_STATUS_SUCCESS;
 }
 
+cusparseStatus_t cusparseLtGetVersion(...) {
+    return CUSPARSE_STATUS_SUCCESS;
+}
+
+cusparseStatus_t cusparseLtGetProperty(...) {
+    return CUSPARSE_STATUS_SUCCESS;
+}
+
 cusparseStatus_t cusparseLtDenseDescriptorInit(...) {
     return CUSPARSE_STATUS_SUCCESS;
 }
